File: client/src/services/researchGroupService.js
import api from "./api";

const researchGroupService = {
  createGroup: async (groupData) => {
    return await api.post("/research-groups/create", groupData);
  },

  getAllGroups: async (
    keyword = "",
    status = "",
    type = "",
    page = 0,
    size = 10
  ) => {
    const params = {
      keyword,
      status,
      type,
      page,
      size,
      sortBy: "createdAt",
      sortDir: "DESC",
    };

    const response = await api.get("/admin/research-groups", {
      params,
    });
    return response;
  },

  getAllGroupPublic: async (
    keyword = "",
    type = "",
    page = 0,
    size = 10
  ) => {
    const params = {
      keyword,
      type,
      page,
      size,
      sortBy: "createdAt",
      sortDir: "DESC",
    };

    const response = await api.get("/research-groups", {
      params,
    });
    return response;
  },

  getGroupById: async (groupId) => {
    const response = await api.get(`/research-groups/${groupId}`);
    return response;
  },

  getMyGroups: async () => {
    const response = await api.get("/research-groups/my-groups");
    return response;
  },

  updateGroup: async (groupId, groupData) => {
    const response = await api.put(`/research-groups/${groupId}`, groupData);
    return response;
  },

  addMember: async (groupId, memberId) => {
    const response = await api.post(
      `/research-groups/${groupId}/members/${memberId}`
    );
    return response;
  },

  removeMember: async (groupId, memberId) => {
    const response = await api.delete(
      `/research-groups/${groupId}/members/${memberId}`
    );
    return response;
  },

  getStatistics: async () => {
    return await api.get("/admin/research-groups/statistics");
  },

  // Admin APIs
  approveGroup: async (groupId) => {
    return await api.put(`/admin/research-groups/${groupId}/approve`);
  },

  rejectGroup: async (groupId, reason) => {
    const response = await api.put(
      `/admin/research-groups/${groupId}/reject`,
      null,
      {
        params: { reason },
      }
    );
    return response;
  },

  deleteGroup: async (groupId) => {
    const response = await api.delete(`/admin/research-groups/${groupId}`);
    return response;
  },

  // Update Google Sheet link
  updateGoogleSheetLink: async (groupId, googleSheetLink) => {
    const response = await api.put(
      `/research-groups/${groupId}/google-sheet-link`,
      { link: googleSheetLink }
    );
    return response.data;
  },

  // Update member info (role and participation rate)
  updateMemberInfo: async (groupId, memberId, role, participationRate) => {
    const response = await api.put(
      `/research-groups/${groupId}/members/${memberId}`,
      { role, participationRate }
    );
    return response.data;
  },

  // Import members from Excel
  importMembersFromExcel: async (groupId, file) => {
    const formData = new FormData();
    formData.append("file", file);
    const response = await api.post(
      `/research-groups/${groupId}/members/import`,
      formData,
      {
        headers: {
          "Content-Type": "multipart/form-data",
        },
      }
    );
    return response;
  },

  // Download Excel template for importing members
  downloadMemberImportTemplate: async () => {
    return await api.get("/research-groups/export-template", {
      responseType: "blob",
    });
  },

  // Document management
  getDocuments: async (groupId) => {
    const response = await api.get(`/research-groups/${groupId}/documents`);
    return response.data;
  },

  createDocument: async (
    groupId,
    documentName,
    documentType,
    description,
    file
  ) => {
    const formData = new FormData();
    formData.append("documentName", documentName);
    formData.append("documentType", documentType);
    if (description) formData.append("description", description);
    formData.append("file", file);
    const response = await api.post(
      `/research-groups/${groupId}/documents`,
      formData,
      {
        headers: {
          "Content-Type": "multipart/form-data",
        },
      }
    );
    return response.data;
  },

  updateDocument: async (
    groupId,
    documentId,
    documentName,
    documentType,
    description,
    file
  ) => {
    const formData = new FormData();
    if (documentName) formData.append("documentName", documentName);
    if (documentType) formData.append("documentType", documentType);
    if (description) formData.append("description", description);
    if (file) formData.append("file", file);
    const response = await api.put(
      `/research-groups/${groupId}/documents/${documentId}`,
      formData,
      {
        headers: {
          "Content-Type": "multipart/form-data",
        },
      }
    );
    return response.data;
  },

  deleteDocument: async (groupId, documentId) => {
    const response = await api.delete(
      `/research-groups/${groupId}/documents/${documentId}`
    );
    return response.data;
  },
};

export default researchGroupService;

File: server/src/main/java/com/example/server/service/ResearchGroupService.java
package com.example.server.service;

import com.example.server.DTO.request.CreateResearchGroupDocumentRequest;
import com.example.server.DTO.request.CreateResearchGroupRequest;
import com.example.server.DTO.request.UpdateResearchGroupRequest;
import com.example.server.DTO.response.ResearchGroupDocumentDTO;
import com.example.server.DTO.response.ResearchGroupJoinRequestDTO;
import com.example.server.DTO.response.ResearchGroupDTO;
import com.example.server.domain.ResearchGroup;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.util.List;

public interface ResearchGroupService {

    // User tạo nhóm (chờ duyệt)
    ResearchGroupDTO createGroup(Integer userId, CreateResearchGroupRequest request);

    // Admin duyệt nhóm (chỉ áp dụng nhóm đang PENDING_ADMIN hoặc PENDING)
    ResearchGroupDTO approveGroup(Integer groupId);

    // Admin từ chối nhóm
    ResearchGroupDTO rejectGroup(Integer groupId, String reason);

    // Giảng viên hướng dẫn duyệt nhóm SV (PENDING_ADVISOR → PENDING_ADMIN)
    ResearchGroupDTO advisorApproveGroup(Integer groupId, Integer advisorId);

    // Giảng viên từ chối nhóm SV (PENDING_ADVISOR → REJECTED)
    ResearchGroupDTO advisorRejectGroup(Integer groupId, Integer advisorId, String reason);

    // Giảng viên lấy danh sách nhóm cần duyệt
    List<ResearchGroupDTO> getGroupsForAdvisor(Integer advisorId, String status);

    // Cập nhật thông tin nhóm (Leader hoặc Admin)
    ResearchGroupDTO updateGroup(Integer groupId, Integer userId, UpdateResearchGroupRequest request);

    // Xóa nhóm (Admin)
    void deleteGroup(Integer groupId);

    // Thêm thành viên (Leader hoặc Admin)
    ResearchGroupDTO addMember(Integer groupId, Integer userId, Integer memberId);

    // Xóa thành viên (Leader hoặc Admin)
    ResearchGroupDTO removeMember(Integer groupId, Integer userId, Integer memberId);

    // Lấy danh sách nhóm (có phân trang, filter)
    Page<ResearchGroupDTO> getAllGroups(String keyword, String status, String type, java.util.Date startDate, java.util.Date endDate, Pageable pageable);

    // Lấy chi tiết nhóm
    ResearchGroupDTO getGroupById(Integer groupId);

    // Lấy danh sách nhóm của user
    List<ResearchGroupDTO> getGroupsByUser(Integer userId);

    // Lấy thống kê
    GroupStatistics getStatistics();

    // Cập nhật Google Sheet link (chỉ thành viên nhóm mới có quyền)
    ResearchGroupDTO updateGoogleSheetLink(Integer groupId, Integer userId, String googleSheetLink);

    // Cập nhật thông tin thành viên (role và participation rate)
    ResearchGroupDTO updateMemberInfo(Integer groupId, Integer userId, Integer memberId, String role,
            Integer participationRate);

    // Import thành viên từ Excel
    ResearchGroupDTO importMembersFromExcel(Integer groupId, Integer userId,
       MultipartFile file) throws IOException;

    byte[] exportTemplate();

    // Document management methods
    List<ResearchGroupDocumentDTO> getDocumentsByGroupId(Integer groupId, Integer userId);
    ResearchGroupDocumentDTO createDocument(Integer groupId, Integer userId, CreateResearchGroupDocumentRequest request, MultipartFile file);
    ResearchGroupDocumentDTO updateDocument(Integer groupId, Integer documentId, Integer userId, CreateResearchGroupDocumentRequest request, MultipartFile file);
    void deleteDocument(Integer groupId, Integer documentId, Integer userId);

    // Đăng ký tham gia nhóm (chờ trưởng nhóm duyệt)
    ResearchGroupJoinRequestDTO requestJoinGroup(Integer groupId, Integer userId, String message);

    List<ResearchGroupJoinRequestDTO> getPendingJoinRequests(Integer groupId, Integer leaderUserId);

    ResearchGroupJoinRequestDTO approveJoinRequest(Long requestId, Integer leaderUserId);

    ResearchGroupJoinRequestDTO rejectJoinRequest(Long requestId, Integer leaderUserId, String reason);

    List<ResearchGroupJoinRequestDTO> getMyJoinRequests(Integer userId);

    class GroupStatistics {
        public long totalGroups;
        public long pendingGroups;
        public long approvedGroups;
        public long rejectedGroups;
    }
}
